extensions/extension.go

```diff
@@ -174,10 +202,36 @@ func evalFunc(env any, name string, args []object.Object) object.Object {
 	return res
 }
 
+// Normalizes to alphanum.gr.
+func sanitizeFileName(file string) (string, error) {
+	if emptyOnly && file != "" {
+		return "", fmt.Errorf("empty only mode, filename must be empty or no arguments, got: %q", file)
+	}
+	if unrestrictedIOs {
+		log.Infof("Unrestricted IOs, not sanitizing filename: %s", file)
+		return file, nil
+	}
+	// only alhpanumeric and _ allowed. no dots, slashes, etc.
+	f := strings.TrimSuffix(file, ".gr")
+	for _, r := range []byte(f) {
+		if !lexer.IsAlphaNum(r) {
+			return "", fmt.Errorf("invalid character in filename %q: %c", file, r)
+		}
+	}
+	return f + ".gr", nil
+}
+
 func saveFunc(env any, _ string, args []object.Object) object.Object {
 	eval := env.(*eval.State)
-	file := args[0].(object.String).Value
-	// Open file for writing.
+	file := ".gr"
+	if len(args) != 0 {
+		file = args[0].(object.String).Value
+		var err error
+		file, err = sanitizeFileName(file)
+		if err != nil {
+			return object.Error{Value: err.Error()}
+		}
+	}
 	f, err := os.Create(file)
```

Comment:
So it could will create a file named .gr if there is no art?

Reply:
That's what "EmptyOnly" is for, allows only 1 file called .gr